### ISSUE TITLE
boost-bootstrap.patch is annoying

### ISSUE THREAD

**Issue by StephanTLavavej** [StephanTLavavej/mingw-distro#2]
Boost's `bootstrap.sh` doesn't recognize mingw-w64 via MSYS2. (mingw.org via MSYS was equally affected.) As I recall, the problem is that Boost's build system has two halves (`bootstrap.sh` and `b2`) which disagree about whether this should be called `mingw` or `gcc`. I was unable to fix this via command-line options, and I tried all possible permutations. My `boost-bootstrap.patch` works around the problem, but it should be fixed properly upstream.

Note that `boost-bootstrap.patch` cannot simply be committed upstream - it would break all other platforms.

**Comment by StephanTLavavej** [comment]
I should consider running cmd /c "bootstrap.bat gcc".

**Comment by StephanTLavavej** [comment]
Build environment instructions:

1. Follow the instructions at http://nuwen.net/mingw.html to install MinGW to C:\MinGW.
2. Verify that you followed "I recommend saying C:\ (as in the screenshot) so it'll create C:\MinGW ." so the directory structure is correct - i.e. you have C:\MinGW\bin (good) instead of C:\MinGW\MinGW\bin (bad).
3. Follow the instructions at https://github.com/StephanTLavavej/mingw-distro/blob/master/README.md to install MSYS2 to C:\Temp\gcc\msys64.
3a. Note that the distro contains 7z, so you can use that to extract .tar.xz.
3b. However, you should be launching MSYS2 from Windows Explorer, not a distro window.
4. Copy the distro's build scripts from C:\MinGW\scripts-13.3 to C:\Temp\gcc\sources-13.3. (C:\Temp is my "scratch directory" and the build scripts assume this path; it is not related to your actual temp directory in Users.)
5. Download boost_1_59_0.tar.bz2, extract it once to boost_1_59_0.tar, and put that in C:\Temp\gcc\sources-13.3.

Now you are ready to build Boost. Here's how:

1. Launch MSYS2 (by clicking msys2_shell.bat in Windows Explorer as usual).
2. `cd /c/temp/gcc/sources-13.3`
3. Copy-paste lines from boost.sh into MSYS2 **one at a time**, inspecting the output carefully. `source ./0_append_distro_path.sh` begins by putting the distro on the path (as MSYS2 doesn't know anything about the distro to start with).
4. If you skip `patch -d /c/temp/gcc/boost_1_59_0 -p1 < boost-bootstrap.patch`, you'll see how the build fails.

Neither the distro nor MSYS2 modify your system permanently. At any time, you can delete C:\MinGW and C:\Temp\gcc\msys64, and they will be gone forever.

**Comment by vprus** [comment]
Steven,
I don't seem to be able to get past bootstrap at all:

```
builtins.c:57:23: fatal error: sys/wait.h: No such file or directory
compilation terminated.
modules/path.c: In function 'path_exists':
modules/path.c:16:12: warning: implicit declaration of function 'file_query' [-Wimplicit-function-declaration]
     return file_query( list_front( lol_get( frame->args, 0 ) ) ) ?
            ^
execunix.c:20:26: fatal error: sys/resource.h: No such file or directory
compilation terminated.
fileunix.c:100:17: fatal error: ar.h: No such file or directory
compilation terminated.
```

This is today's develop branch of Boost. Am I doing something wrong, or it's a real problem?

Also, relatedly, can you clarify whether step 6 in your instructions is supposed to create "C:/Temp/gcc/sources-13.4/scripts-13.4", or the content of the 'scripts-13.4' directory should be directly present under "C:/Temp/gcc/sources-13.4". It might not matter in the end, if I reproduce the problem directly, but still.

Thanks,
Volodya

**Comment by StephanTLavavej** [comment]
> builtins.c:57:23: fatal error: sys/wait.h: No such file or directory

Looks like bootstrap.sh is detecting the system as a Unix. Are you applying my boost-bootstrap.patch, which forces "mingw" to be passed to build.sh? If so, you should be able to successfully bootstrap. If you aren't, then you're reproing the problem - bootstrap.sh won't succeed by itself. (Nor have I found any combination of command-line parameters to bootstrap.sh followed by b2 that allow them to both succeed without patching).

> or the content of the 'scripts-13.4' directory should be directly present under "C:/Temp/gcc/sources-13.4".

That's the one you should have. The subdirectory's name doesn't matter, it just needs to be C:\Temp\gcc\SOMETHING which contains both the various build scripts, and the source tar files.

**Comment by vprus** [comment]
Okay, I see bootstrap succeed with './bootstrap.sh --with-toolset=mingw', so yes, looks like I've reproduced at least one issue. Thanks!

**Comment by StephanTLavavej** [comment]
Awesome, thanks for investigating! Let me know if there's anything else I can do to help.

**Comment by vprus** [comment]
I believe this should be fixed in https://github.com/boostorg/build/commit/2f8e4ae8681fba1e2814a305c130751e7492e68a - there is no longer a bootstrap toolchain called 'mingw', the decision to build either posix- or windows-specific sources is made by checking `gcc -dumpmachine`. I could bootstrap and build a library from the Boost root inside minsys shell with no issues whatsoever. Please let me know whether it works on your side.

**Comment by StephanTLavavej** [comment]
Yep, your patch works for me. Thank you so much! Will this ship in Boost 1.61?

**Comment by vprus** [comment]
Great, and yes, this patch is included in master branch and should make it to 1.61.
